math/src/test/kotlin/ktx/math/ImmutableVector2Test.kt

```diff
@@ -0,0 +1,712 @@
+        vectors.forEach { v1 ->
+            vectors.forEach { v2 ->
+                assertEquals(v1.toMutable().isPerpendicular(v2.toMutable()), v1.isPerpendicular(v2))
+                scalars.forEach { e ->
+                    assertEquals(v1.toMutable().isPerpendicular(v2.toMutable(), e), v1.isPerpendicular(v2, e))
+                }
+            }
+        }
+    }
+
+    @Test
+    fun `hasSameDirection should return same than Vector2`() {
+        vectors.forEach { v1 ->
+            vectors.forEach { v2 ->
+                assertEquals(v1.toMutable().hasSameDirection(v2.toMutable()), v1.hasSameDirection(v2))
+            }
+        }
+    }
+
+    @Test
+    fun `hasOppositeDirection should return same than Vector2`() {
+        vectors.forEach { v1 ->
+            vectors.forEach { v2 ->
+                assertEquals(v1.toMutable().hasOppositeDirection(v2.toMutable()), v1.hasOppositeDirection(v2))
+            }
+        }
+    }
+
+    @Test
+    fun `copy should return vector with same properties`() {
+        vectors.forEach { vector ->
+            assertEquals(vector.x, vector.copy(y = 42f).x)
+            assertEquals(42f, vector.copy(y = 42f).y)
+            assertEquals(42f, vector.copy(x = 42f).x)
+            assertEquals(vector.y, vector.copy(x = 42f).y)
+            assertEquals(ImmutableVector2(1f, 2f), vector.copy(x = 1f, y = 2f))
+        }
+    }
+
+    @Test
+    fun `should destruct vector into two floats`() {
+        vectors.forEach { vector ->
+            val (x, y) = vector
+
+            assertEquals(vector.x, x)
+            assertEquals(vector.y, y)
+        }
+    }
+
+    @Test
+    fun `should compare vectors by length`() {
+        var previousLength = -1f
+        var previousLength2 = -1f
+
+        vectors.toList().shuffled().sorted().forEach { vector ->
+            assertTrue(vector.len >= previousLength)
+            assertTrue(vector.len2 >= previousLength2)
+            previousLength = vector.len
+            previousLength2 = vector.len2
+        }
+    }
+
+
+    @Test
+    fun `should increment vector values with ++ operator`() {
+        var vector = ImmutableVector2(10f, 10f)
+
+        vector++
+
+        assertEquals(11f, vector.x, MathUtils.FLOAT_ROUNDING_ERROR)
+        assertEquals(11f, vector.y, MathUtils.FLOAT_ROUNDING_ERROR)
+    }
+
+    @Test
+    fun `should decrement vector values with -- operator`() {
+        var vector = ImmutableVector2(10f, 10f)
+
+        vector--
+
+        assertEquals(9f, vector.x, MathUtils.FLOAT_ROUNDING_ERROR)
+        assertEquals(9f, vector.y, MathUtils.FLOAT_ROUNDING_ERROR)
+    }
+
+    @Test
+    fun `should divide vectors by float scalars with div operator`() {
+        val vector = ImmutableVector2(10f, 10f)
+
+        val result = vector / 2.5f
+
+        assertEquals(4f, result.x, MathUtils.FLOAT_ROUNDING_ERROR)
+        assertEquals(4f, result.y, MathUtils.FLOAT_ROUNDING_ERROR)
+    }
```

Comment:
Most of these tests check if the implementation is the same as the official one, but not if it is correct. I'm not saying there are any bugs in the code, but if there were present in both our and LibGDX code, these tests wouldn't expose that at all. You've already done a ton of work on this PR, and this is not a pressing issue, but if you don't mind adding a few tests that check the results of `ImmutableVector2` methods against actual pre-calculated values that we know are correct, I'd be much more confident before publishing this code. You don't have to do it for every element of `vectors` though.

Reply:
This is a fair concern and I share it, especially after finding incorrect behavior in LibGDX's vector.

And because some implementation also delegate to LibGDX, a bug in LibGDX would necessirly be present and not detected.

I'll try to add some check against actual pre-calculated values.

Reply:
Thanks, much appreciated.